contrib/postgres_fdw/transcoding.c

```
@@ -0,0 +1,241 @@
+
+	memset(&flinfo, 0, sizeof(FmgrInfo));
+	fmgr_info_cxt(fmgr_internal_function("int8_avg_serialize"), &flinfo, CurrentMemoryContext);
+	Init_AggState(&aggstate);
+	return CallAggfunction1(&flinfo, (Datum)internal_polynum, (fmNodePtr *)&aggstate);
+}
+
+/*
+ * Handle trancoding for aggregate function -- avg(numeric), aggfnoid = 2114
+ * The input str is formatted like {$sum}
+ * 1. Transcode the input str into numeric
+ * 2. Construct internal type NumericAggState
+ * 3. Serialize NumericAggState into bytea
+ */
+static Datum transfn_for_sum_numeric(PG_FUNCTION_ARGS)
+{
+	char		*str = PG_GETARG_CSTRING(0);
+	int32		typmod = PG_GETARG_INT32(2);
+	FmgrInfo	flinfo;
+	Datum		sumd;
+	NumericAggState *target_state;
+	AggState 	aggstate;
+
+	memset(&flinfo, 0, sizeof(FmgrInfo));
+	fmgr_info_cxt(fmgr_internal_function("numeric_in"), &flinfo, CurrentMemoryContext);
+	sumd = InputFunctionCall(&flinfo, str, NUMERICOID, typmod);
+
+	target_state = makeNumericAggStateCurrentContext(false);
+	do_numeric_accum(target_state, (Numeric)sumd);
+	// N(count) is not used for sum(), so here we set N = 1 as default value.
+	target_state->N = 1;
+
+	memset(&flinfo, 0, sizeof(FmgrInfo));
+	fmgr_info_cxt(fmgr_internal_function("numeric_avg_serialize"), &flinfo, CurrentMemoryContext);
+	Init_AggState(&aggstate);
+	return CallAggfunction1(&flinfo, (Datum)target_state, (fmNodePtr *)&aggstate);
+}
+
+PGFunction GetTranscodingFnFromOid(Oid aggfnoid) {
+	PGFunction refnaddr = NULL;
+	if (aggfnoid == InvalidOid) 
+	{
+		return NULL;
+	}
+	switch (aggfnoid) 
+	{
+		case 2100:
+			// - 2100 pg_catalog.avg int8|bigint
+			refnaddr = transfn_for_avg_bigint;
+			break;
+		case 2103:
+			// - 2103 pg_catalog.avg numeric
+			refnaddr = transfn_for_avg_numeric;
+			break;
+		case 2107:
+			// - 2107 pg_catalog.sum int8|bigint
+			refnaddr = transfn_for_sum_bigint;
+			break;
+		case 2114:
+			// - 2114 pg_catalog.sum numeric
+			refnaddr = transfn_for_sum_numeric;
+			break;
+		default:
+			break;
```

For other unsupported aggfnoid, should we raise an ERROR here?

---

For those unsupported aggfnoids, gpdb will generate query plan that it just fetches all data from remote servers then does aggregate in local.

So I think it's not necessary to raise an ERROR here. Or an WARNING instead?

---

If everything will be correct and this will be easy to figure out by reading the plan, then current logic is good enough.

---

sry, I missed, if you want to add some logs then DEBUG5 is enough.